torch/csrc/jit/passes/convert_activation.cpp

```diff
@@ -0,0 +1,100 @@
+#include <torch/csrc/jit/passes/convert_activation.h>
+#include <torch/csrc/jit/passes/remove_mutation.h>
+#include "ATen/core/interned_strings.h"
+
+namespace torch {
+namespace jit {
+
+namespace {
+static const std::unordered_set<Symbol> inplace_activation_ops = {
+    aten::hardsigmoid_,
+    aten::hardtanh_,
+    aten::hardswish_,
+    aten::relu_,
+    aten::relu6_,
+    aten::sigmoid_,
+    aten::tanh_};
+} // namespace
+
+FunctionalActivationRewriter::FunctionalActivationRewriter(
+    std::shared_ptr<Graph> graph)
+    : aliasDb_(nullptr), graph_(std::move(graph)) {
+  for (auto op : inplace_activation_ops) {
+    std::string name = std::string(op.toQualString());
+    name.pop_back();
+    functional_activation_ops_.insert(Symbol::fromQualString(name));
+  }
+}
+
+bool FunctionalActivationRewriter::FunctionalToInplaceActivation(Block* block) {
+  bool changed = false;
+  for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+    auto* node = *it;
+    it++;
+
+    for (Block* sub_block : node->blocks()) {
+      changed |= FunctionalToInplaceActivation(sub_block);
+    }
+
+    if (functional_activation_ops_.count(node->kind()) == 0) {
+      continue;
+    }
+
+    Value* input = node->inputs().at(0);
+    Value* output = node->output();
+
+    // Use y = relu(x) as an example:
+    // If y and x have different type, skip the conversion
+    if (input->type() != output->type()) {
```

Comment:
So, this check is getting at the fact that we can't inplace a node if the output dtype is separate from the input dtype. Could you comment that ? Shape isn't necessary for activation since the size will always remain unchanged. 

If we're going to use `input->type() != output->type()` to verify that the output dtype equals the input dtype, we also need to check that the dtype is actually specified on the types. It is not guaranteed to be. 

I would maintain a list of activations you know do not type promote (relu for instance), and then for the ones not in that list that do type promote (sigmoid), check that the dtype is specified on both input and output and they are equal.